packages/tiktok-video-element/tiktok-video-element.js

```diff
@@ -0,0 +1,302 @@
+    this.render();
+  }
+
+  disconnectedCallback() {
+    window.removeEventListener('message', this._onMessage);
+  }
+
+  attributeChangedCallback(name, oldValue, newValue) {
+    this.render();
+
+    if (name === 'muted') {
+      if (newValue !== null && newValue !== 'false') {
+        this.mute();
+      } else {
+        this.unMute();
+      }
+    }
+  }
+
+  _createPublicPromise() {
+    let res, rej;
+    const p = new Promise((r, j) => {
+      res = r;
+      rej = j;
+    });
+    p.resolve = res;
+    p.reject = rej;
+    return p;
+  }
+
+  get attrs() {
+    return TikTokVideoElement.observedAttributes.reduce((o, attr) => {
+      const val = this.getAttribute(attr);
+      if (val !== null) o[attr] = val === '' ? '1' : val;
+      return o;
+    }, {});
+  }
+
+  async resolveVideoId() {
+    const src = this.getAttribute('src');
+    if (!src) return null;
+
+    // 1. Direct numeric ID
+    if (/^\d{10,}$/.test(src)) return src;
+
+    // 2. player/v1 URLs
+    let m = src.match(/player\/v1\/(\d+)/);
+    if (m) return m[1];
+
+    // 3. embed, share paths, shareId or item_id
+    m = src.match(/(?:embed|v|usr|user)\/(\d+)/) || src.match(/[?&](?:shareId|item_id)=(\d+)/);
+    if (m) return m[1];
+
+    // 4. watch URLs including optional username
+    m = src.match(/\/@[^/]*\/video\/(\d+)/);
+    if (m) return m[1];
+
+    // 5. /@/video/ format
+    m = src.match(/\/@\/video\/(\d+)/);
+    if (m) return m[1];
+
+    // 6. Short links—resolve via HEAD redirect
+    if (/vm\.tiktok\.com|vt\.tiktok\.com/.test(src)) {
+      try {
+        const resp = await fetch(src, { method: 'HEAD', redirect: 'follow' });
+        const final = resp.url;
+        m =
+          final.match(/\/video\/(\d+)/) ||
+          final.match(/player\/v1\/(\d+)/) ||
+          final.match(/[?&](?:shareId|item_id)=(\d+)/);
+        if (m) return m[1];
+      } catch (e) {
+        console.error('Error resolving short link', e);
+      }
+    }
+
+    return null;
+  }
+
+  async render() {
+    const id = await this.resolveVideoId();
+    if (!id) return;
+
+    this.loadComplete = this._createPublicPromise();
+
+    const params = {
+      controls: this.attrs.controls === undefined ? null : 1,
+      autoplay: this.attrs.autoplay,
+      loop: this.attrs.loop,
+      rel: 0,
+      ...this.config,
+    };
+
+    const src = `${TikTokVideoElement.EMBED_BASE}/${id}?${this.serialize(params)}`;
+    const dataConfigAttr = this.config ? `data-config='${JSON.stringify(this.config).replace(/'/g, '&apos;')}'` : '';
+
+    this.shadowRoot.innerHTML = `
+      <style>
+        :host { display:inline-block;position:relative;width:100%;height:100%; }
+        iframe { position:absolute;top:0;left:0;width:100%;height:100%;border:0;}
+      </style>
+      <iframe src="${src}" ${dataConfigAttr} allow="autoplay; fullscreen" title="TikTok video"></iframe>
+    `;
+
+    this.iframe = this.shadowRoot.querySelector('iframe');
+
```

Comment:
all the media elements can be server side rendered via the React wrapper. for the ones with iframes we also want to SSR the iframe because it means the iframe can be loaded before any JS since it's already in the html page.

please have a look at the Vimeo or Youtube element to see how it is done.